code/game/machinery/vending/marine_vending.dm

```diff
@@ -238,6 +243,112 @@
+					/obj/item/cell/lasgun/M43 = 10,
+					/obj/item/cell/lasgun/M43/highcap = 2,
+					)
+
+	contraband =   list()
+
+	premium = list()
+
+	prices = list()
+
+/obj/machinery/vending/lasgun/New()
+	..()
+	update_icon()
+
+/obj/machinery/vending/lasgun/update_icon()
+	if(machine_max_charge)
+		switch(machine_current_charge / machine_max_charge)
+			if(0)
+				icon_state = "lascharger-off"
+			if(1 to 0.76)
+				icon_state = "lascharger"
+			if(0.75 to 0.51)
+				icon_state = "lascharger_75"
+			if(0.50 to 0.26)
+				icon_state = "lascharger_50"
+			if(0.25 to 0.01)
+				icon_state = "lascharger_25"
+
+/obj/machinery/vending/lasgun/attackby(obj/item/W, mob/user)
+	if(istype(W, /obj/item/cell/lasgun))
+		stock(W, user, TRUE)
+		return TRUE
+	return ..()
+
+/obj/machinery/vending/lasgun/examine(mob/user)
+	. = ..()
+	to_chat(user, "<b>It has [machine_current_charge] of [machine_max_charge] charge remaining.</b>")
+
+
+/obj/machinery/vending/lasgun/MouseDrop_T(var/atom/movable/A, mob/user)
+	. = ..()
+
+	var/obj/item/I = A
+	if(istype(I, /obj/item/cell/lasgun))
+		stock(I, user, TRUE)
+	else
+		stock(I, user)
+
+/obj/machinery/vending/lasgun/stock(obj/item/item_to_stock, mob/user, recharge = FALSE)
+	var/datum/data/vending_product/R //Let's try with a new datum.
+	 //More accurate comparison between absolute paths.
+	for(R in (product_records + hidden_records + coin_records))
+		if(item_to_stock.type == R.product_path && !istype(item_to_stock,/obj/item/storage)) //Nice try, specialists/engis
+			if(istype(item_to_stock, /obj/item/cell/lasgun) && recharge)
+				if(!recharge_lasguncell(item_to_stock, user))
+					return //Can't recharge so cancel out
+
+			if(item_to_stock.loc == user) //Inside the mob's inventory
+				if(item_to_stock.flags_item & WIELDED)
+					item_to_stock.unwield(user)
+				user.temp_drop_inv_item(item_to_stock)
+
+			if(istype(item_to_stock.loc, /obj/item/storage)) //inside a storage item
+				var/obj/item/storage/S = item_to_stock.loc
+				S.remove_from_storage(item_to_stock, user.loc)
+
+			qdel(item_to_stock)
+			if(!recharge)
+				user.visible_message("<span class='notice'>[user] stocks [src] with \a [R.product_name].</span>",
+				"<span class='notice'>You stock [src] with \a [R.product_name].</span>")
+			R.amount++
+			updateUsrDialog()
+			break //We found our item, no reason to go on.
+
+/obj/machinery/vending/lasgun/proc/recharge_lasguncell(obj/item/cell/lasgun/M43/A, mob/user)
+	var/recharge_cost = (A.maxcharge - A.charge)
+	if(recharge_cost > machine_current_charge)
+		to_chat(user, "<span class='warning'>[A] cannot be recharged; [src] has inadequate charge remaining: [machine_current_charge] of [machine_max_charge].</span>")
+		return
+	else
+		to_chat(user, "<span class='warning'>You insert [A] into [src] to be recharged.</span>")
+		if(icon_vend)
+			flick(icon_vend,src)
+		playsound(loc, 'sound/machines/hydraulics_1.ogg', 25, 0, 1)
+		machine_current_charge -= min(machine_current_charge, recharge_cost)
```

Comment:
You remove the charge correctly but it is not added into the cell itself.
You do not call the ..() proc that actually deletes the item to insert if that's your intention, otherwise you should just add charge into the cell.

Reply:
This is handled